BUG: Fix pickle and memoryview for datetime64, timedelta64 scalars

mattip opened the pull request:
Fixes #11656, which is a regression after #10564.

Adds tests for pickling datetime64 and timedelta64 scalars. Pickling arrays of these `dtype`s works but pickling scalars fails. 

The problem seems to be in the new code for `gentype_getbuffer`, which creates a PEP 3118 compliant buffer. The old code was:
```C
int gentype_getbuffer(PyObject *self, Py_buffer *view, int flags)
{
    Py_ssize_t len;
    void *buf;
    /* FIXME: XXX: the format is not implemented! -- this needs more work */
    len = gentype_getreadbuf(self, 0, &buf);
    return PyBuffer_FillInfo(view, self, buf, len, 1, flags);
}
```
so now I need to find what is happening in that code that is not happening in the new code.

So far I only found where we override an original exception with a new one. ~Now `memoryview(np.datetime64('2016-01-01'))` raises an exception with the same message as `memoryview(np.array(100, 'M8[D]').~

Edit: python2 - python3 confusion

mattip committed:
MAINT: test memoryview, improve error message, don't swallow first exception

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -796,8 +796,10 @@ gentype_getbuffer(PyObject *self, Py_buffer *view, int flags)
     info = _buffer_get_info(self);
     if (info == NULL) {
-        PyErr_SetString(PyExc_BufferError,
+        if (!PyErr_Occurred()) {
```
I don't think this `if` can ever be hit.

mattip replied:
Try setting a breakpoint there and calling `memoryview(np.datetime64('2016-01-01'))` or more to the point of the issue `pickle.dumps(np.datetime64('2016-01-01'))`. The call chain is `gettype_getbuffer` -> `_buffer_get_info` -> `_buffer_info_new` -> `_buffer_format_string` -> switch statement -> missing case marked with XXX -> raising error with "cannot include dtype '%c' in a buffer".

eric-wieser replied:
To be clear - I think `_buffer_get_info` is well behaved, and will always set an exception if it returns `NULL`.

mattip replied:
ok

eric-wieser commented:
In case it helps, this was the old behavior
```
>>> memoryview(np.datetime64('2016-01-01')).format
'B'
```

mattip committed:
BUG: add tests, special case memoryview(datetime64-scalar)

mattip commented:
The memoryview info for `datetime64` and `timedelta64` uses `B` which is a single byte, so the code needs to special case the memoryview buffer generation. The `dtype.itemsize == 8`, `dtype.shape == ()` becomes the memoryview `mv.itemsize == 1`, `mv.shape == (8,0)`. If the dtype used `L' or 'l',  the code would be cleaner.

mattip commented:
I cleared the WIP in the title, this seems to work now.

eric-wieser commented on numpy/core/tests/test_regression.py:
```diff
@@ -2412,3 +2412,27 @@ def test_field_access_by_title(self):
+        dt1 = np.datetime64('2016-01-01')
+        dt2 = np.datetime64('2017-01-01')
+        expected = {'strides': (1,), 'itemsize': 1, 'ndim': 1,
+                    'shape': (8,), 'format': 'B'}
```
One interesting idea would be to set the format to `T{:value:8B  :dtype_m8[N]:(0)B}` so that we can actually round-trip this

eric-wieser commented on numpy/core/tests/test_regression.py:
```diff
@@ -2412,3 +2412,27 @@ def test_field_access_by_title(self):
+        assert_equal(res, expected)
+        assert_equal(v.tolist(), lst)
+
+        lst = [110, 1, 0, 0, 0, 0, 0, 0]
```
I think this might be making assumptions about endianness

mattip replied:
removed

eric-wieser commented on numpy/core/tests/test_regression.py:
```diff
@@ -2412,3 +2412,27 @@ def test_field_access_by_title(self):
                 assert_(base <= sys.getrefcount(s))
+
+    @pytest.mark.skipif(sys.version_info < (3, 0), reason="Python3 only")
+    def test_datetime_memoryview(self):
```
I think this test might be better alongside wherever the other memoryview tests are

mattip replied:
moved

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -363,7 +363,10 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
-        case NPY_UBYTE:        if (_append_char(str, 'B')) return -1; break;
+        case NPY_UBYTE:
+        case NPY_DATETIME:
+        case NPY_TIMEDELTA:    if (_append_char(str, 'B')) return -1; break;
```
Would it be legal to set the format to `8B` instead, and leave the shape as `()`?

mattip replied:
Why not 'Q' then, since we call these `datetime64` and `timedelta64`, the 64 bit is explicit. This would mean that `d=np.ones((3), 'M8[D]'); np.array(memoryview(d))` would return a `uint64` array, rather that a 3 by 8 'uint8' one.

eric-wieser commented:
Can you add a test for `memoryview` of datetime arrays?

mattip commented:
> Can you add a test for memoryview of datetime arrays?

Added tests for both scalars and arrays. After changing the format string to 'Q' I could remove the special casing in `memoryview(scalar)` added in the earlier commits. Also now `np.asarray(memoryview(d))` works where it used to raise an error (also improved the error message).

I tried to hit the "Sanity check" errors in tests. Should I go to C to create non-compliant buffers?

mattip commented:
Would this be easier to review if I move the last commit 63e44e3 to a separate PR? Then we could get this into 1.15.1 without changing the buffer format from `B`. Note various things would fail, but they failed previously (all on python3) where 
```
m=memoryview(np.array((3), 'M8[D]'))
d = memoryview(np.datetime64('2001-01-01'))
a = np.array(m)
b = np.array(d)
```
version | m succeeds? | d succeeds?  | a succeeds? | b succeeds?
-----|-----|-----|-----|-----
1.13.0 | no | yes | (not relevant) | yes
1.14.0 | no | yes | (not relevant) | yes
1.15.0 | no | no | (not relevant) | (not relevant)
895a2a8 (format=`B`) | yes | yes | no | yes
63e44e3 (format=`Q`) | yes | yes | yes | yes

charris commented:
Does the 'Q' format need a prefix, '=' perhaps, to indicate standard size (8) and avoid compiler dependence? I'm thinking of https://docs.python.org/2/library/struct.html.

eric-wieser commented:
> Then we could get this into 1.15.1 without changing the buffer format from B

I agree - let's leave the buffer format as it used to be in 1.15

What do you mean by "a doesn't succeed"?

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1390,10 +1389,12 @@ _array_from_buffer_3118(PyObject *memoryview)
-                    "Item size computed from the PEP 3118 buffer format "
-                    "string does not match the actual item size.");
+                    "Item size %d for PEP 3118 buffer format "
+                    "string %s does not match the dtype %c item size %d.",
```
Better to use the `INTP` format macro here (or perhaps `%z` for itemsize, rather than casting.

mattip commented:
I backed out the last commit to fix only the regression on pyhon 3 between 1.14 and 1.15

eric-wieser commented:
What is the exception that's thrown?

mattip commented:
after 895a2a8:
```
python3 -c "import numpy as np; np.array(memoryview(np.array((3), 'M8[D]')))"
```
prints
```
RuntimeError: Item size computed from the PEP 3118 buffer format string \
does not match the actual item size.
```
This was my motivation for improving the error message in 63e44e3.

eric-wieser commented:
Can you show what the error message was _with_ the message improvement? I was hoping the message improvement would make it into 1.15.1, since it's very useful for us debugging.

mattip commented:
It now becomes (line broken for smaller screen formatting)
```
RuntimeError: Item size 8 for PEP 3118 buffer format string B does not match \
the dtype B item size 1.
```
Can we leave this PR at fixing #11656, and if needed open further issues and PRs around memoryviews of datetime arrays (as opposed to scalars, needed for pickling)? I am fine with what we have done so far, but feel we have reached a nice stopping point that will enable 1.15.1 to be released soon.

eric-wieser commented:
If `m` succeeds but `a` doesn't, then it sounds like `m` has gone from failing to silently doing the wrong thing and producing an invalid buffer.

eric-wieser commented:
A more worrying and related regression on master (and possibly 1.15):

```
>>> dt = np.dtype([('a', int), ('b', 'M8[s]')])
>>> a = np.empty(1, dt)
>>> memoryview(a[0])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
ValueError: cannot include dtype 'M' in a buffer

During handling of the above exception, another exception occurred:

SystemError                               Traceback (most recent call last)
<ipython-input-17-d9b2a1f0b24c> in <module>()
----> 1 memoryview(a[0])

SystemError: <class 'memoryview'> returned a result with an error set
```

mattip commented:
That example succeeds on this branch
```
>>> def as_dict(m):
...    return dict(shape=m.shape, itemsize=m.itemsize, ndim=m.ndim, format=m.format)
>>>
>>> m = memoryview(a[0])
>>> as_dict(m)
{'shape': (), 'itemsize': 16, 'ndim': 0, 'format': 'T{l:a:B:b:}'}
```

eric-wieser commented:
Well, except that format contradicts the itemsize, which is only possibly worse than a `SystemError`

mattip commented:
Found the reason for the `SystemError`, see #11754.

charris commented:
@eric-wieser Good to go?

mattip commented:
@eric-wieser I think we could trap `memoryview(array)` that creates non-compliant PEP-3118 buffers in `array_getbuffer`(at creation), which would preserve the previous failing behaviour. Something like
```
PyArray_Descr descr = _descriptor_from_pep3118_format(view->format);
if (descr->elsize != view->itemsize) {
    /* raise BufferError like previously */
}
```

It would need a test for structured arrays.

mattip commented:
once #11754 lands, I can merge it and expande the added test there to make sure we are not creating non-compliant PEP 3118 buffers

mattip commented:
rebased and reworked. Now the test for failure added in #11754 passes without failure, so I reshuffled the tests. Also added a check for PEP 3118 compliance in `array_getbuffer`, so now `memoryview(np.array((3), 'M8[D]'))` fails as it did until the first commits of this branch.

Once this lands and the dust settles around 1.15.1, we could revisit PEP compliant buffers of datetime objects, either through #11749 or a replacement.

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -6487,6 +6487,22 @@ def test_roundtrip_scalar(self):
+        dt_tag = np.dtype({'names':['a','b'], 'formats':['<i8','u1'],
+                           'offsets':[0,8], 'itemsize':16})
+        a = np.empty(3, dt)
+        m = memoryview(a)
```
I'd like to see an explicit `as_dict` test here to be sure that the buffer is actually valid

eric-wieser commented on numpy/core/tests/test_scalarbuffer.py:
```diff
@@ -78,8 +78,29 @@ def test_void_scalar_structured_data(self):
-        assert_raises((ValueError, BufferError), memoryview, a[0])
+        m = memoryview(a[0])
+        assert_equal(m.format, 'T{l:a:B:b:}')
+        assert_equal(m.itemsize, 16)
```
Taken together, these two lines violate PEP3118. If we can't produce a valid format string, we should just error.

mattip replied:
Using `'uint16'` instead of `int` in the dtype would make the `itemsize=16` valid, correct?

mattip replied:
ahh, no sorry, I see the problem.

eric-wieser commented on numpy/core/tests/test_scalarbuffer.py:
```diff
@@ -78,8 +78,29 @@ def test_void_scalar_structured_data(self):
         a = np.empty(1, dt)
-        assert_raises((ValueError, BufferError), memoryview, a[0])
+        m = memoryview(a[0])
+        assert_equal(m.format, 'T{l:a:B:b:}')
```
Safe to test with `uint16` which doesn't have the platform-dependent width.

mattip commented:
added test to `gentype_getbuffer` for PEP 3118 compliant buffer creation, test now raises instead of passing (as it should)

charris commented:
@eric-wieser  Look good?

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -833,13 +862,34 @@ gentype_getbuffer(PyObject *self, Py_buffer *view, int flags)
 #endif
     view->len = elsize;
+    if (PyArray_IsScalar(self, Datetime) || PyArray_IsScalar(self, Timedelta)) {
+        elsize = 1; /* descr->elsize is 8, but scalars are always 1 */
```
This comment should be explaining that the format is being exposed as `B` - `scalars are always 1` is false - `sizeof(B) == 1` is what's relevant here.

mattip replied:
removing code, comment expanded above.

mattip replied:
Whoops, this is needed. Comment fixed.

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -833,13 +862,34 @@ gentype_getbuffer(PyObject *self, Py_buffer *view, int flags)
     view->suboffsets = NULL;
     view->obj = self;
+    if (view->format) {
+        /* XXX remove me, this should never occur */
```
I assume that in the tests this does occur? If so, perhaps say when it occurs in the comment.

mattip commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -483,7 +484,18 @@ _buffer_info_new(PyObject *obj)
+        info->shape = malloc(sizeof(Py_ssize_t));
+        info->strides = malloc(sizeof(Py_ssize_t));
+        info->shape[0] = 8;
+        info->strides[0] = 1;
```
Good idea

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -483,7 +483,25 @@ _buffer_info_new(PyObject *obj)
+        _append_char(&fmt, 'B');
+        _append_char(&fmt, '\0');
+        info->shape[0] = 8;
+        info->strides[0] = 1;
```
I think `shape` and `strides` are unitialized here - they're malloc'd below for some cases.

mattip commented:
Special-cased `memoryview(d)` where `d` is a `datetime64` scalar. This now fixes the original issue, adds tests for `memoryview` and `pickle` ing `datetime64` objects, and never creates non-compliant buffers.

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -483,7 +483,25 @@ _buffer_info_new(PyObject *obj)
+         */
+        _append_char(&fmt, 'B');
+        _append_char(&fmt, '\0');
+        info->shape = malloc(sizeof(Py_ssize_t) * 2);
```
malloc can fail - this needs to call `PyErr_NoMemory()` if so

mattip replied:
fixed

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -483,7 +483,25 @@ _buffer_info_new(PyObject *obj)
 
-    if (PyArray_IsScalar(obj, Generic)) {
+    if (PyArray_IsScalar(obj, Datetime) || PyArray_IsScalar(obj, Timedelta)) {
+        info->ndim = 1;
```
I'd prefer this appear after the comment, alongside all the other assigments

mattip replied:
fixed

eric-wieser reviewed (approved):
Looks good now, thanks for the persistence. Will merge and squash if you don't want to squash yourself.

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -483,7 +483,29 @@ _buffer_info_new(PyObject *obj)
+         * in _buffer_format_string.
+         */
+        _append_char(&fmt, 'B');
+        _append_char(&fmt, '\0');
```
Something for another PR: We ought to be checking the return value of `_append_char`, but we don't

mattip commented:
sqaushed to two commits.
